internal/examples/agent/agent/agent.go

```diff
@@ -356,17 +371,96 @@ func (agent *Agent) Shutdown() {
 	}
 }
 
+// requestClientCertificate sets a request to be sent to the Server to create
+// a client certificate that the Agent can use in subsequent OpAMP connections.
+// This is the initiating step of the Client Signing Request (CSR) flow.
+func (agent *Agent) requestClientCertificate() {
+	if agent.certRequested {
+		// Request only once, for bootstrapping.
+		// TODO: the Agent may also for example check that the current certificate
+		// is approaching expiration date and re-requests a new certificate.
+		return
+	}
+
+	// Generate a keypair for new client cert.
+	clientCertKeyPair, err := rsa.GenerateKey(cryptorand.Reader, 4096)
+	if err != nil {
+		agent.logger.Errorf("Cannot generate keypair: %v", err)
+		return
+	}
+
+	// Encode the private key of the keypair as DER.
+	privateKeyDER := x509.MarshalPKCS1PrivateKey(clientCertKeyPair)
+
+	// Convert private key from DER to PEM.
+	privateKeyPEM := new(bytes.Buffer)
+	pem.Encode(
+		privateKeyPEM, &pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: privateKeyDER,
+		},
+	)
+	// Keep it. We will need it in later steps of the flow.
+	agent.clientPrivateKeyPEM = privateKeyPEM.Bytes()
+
+	// Create the CSR.
+	template := x509.CertificateRequest{
+		Subject: pkix.Name{
+			CommonName:   "OpAMP Example Client",
+			Organization: []string{"OpenTelemetry OpAMP Workgroup"},
+			Locality:     []string{"Agent-initiated"},
+			// Where do we put instance_uid?
```

Comment:
It can also part of CommonName. We don't really need to specify it in OpAMP spec. It is up to the Agent and Server to agree what the certificate contains.

Reply:
I think `CommonName` probably makes the most sense, from my understanding it is conventionally considered to be the most specific part of the distinguished name in the certificate.